### PR TITLE
Add support for IPV6_TCLASS and IPV6_RECVTCLASS

### LINKED ISSUE #363: Support IPV6_TCLASS
`IPV6_TCLASS` is `IP_TOS` in IPv6. It is already defined in `libc` ([ref](https://docs.rs/libc/0.2.138/libc/constant.IPV6_TCLASS.html)).

A simple solution would be `pub(crate) use libc::IPV6_TCLASS;`. But I'm not sure how to correctly configure the cfg attribute.

### PR DESCRIPTION
**Pull request by duskmoon314**
A simple implementation of the IPV6_TCLASS and IPV6_RECVTCLASS modeled after the existing IP_TOS and IP_RECVTOS implementations.

close #363

### CODE REVIEW

File: src/socket.rs [hunk 1]
```diff
@@ -1668,6 +1668,98 @@ impl Socket {
...
         }
     }
+
+    /// Set the value of the `IPV6_TCLASS` option for this socket.
+    ///
+    /// Specifies the traffic class field that is used in every packets
+    /// sent from this socket.
+    ///
+    /// NOTE: <https://learn.microsoft.com/en-us/windows/win32/winsock/ipproto-ipv6-socket-options>
+    /// does not list this option, but it is supported in `windows-sys`.
```
Comment (Thomasdezeeuw) [inline src/socket.rs:1678]: Don't need to mention this.
```suggestion
```

File: src/socket.rs [hunk 2]
```diff
@@ -1668,6 +1668,98 @@ impl Socket {
...
+        }
+    }
+
+    /// Get the value of the `IPV6_TCLASS` option for this socket.
+    ///
+    /// For more information about this option, see [`set_tclass_v6`].
+    ///
+    /// NOTE: <https://learn.microsoft.com/en-us/windows/win32/winsock/ipproto-ipv6-socket-options>
+    /// does not list this option, but it is supported in `windows-sys`.
+    ///
```
Comment (Thomasdezeeuw) [inline src/socket.rs:1702]:
```suggestion
```

File: src/socket.rs [hunk 3]
```diff
@@ -1668,6 +1668,98 @@ impl Socket {
...
+        target_os = "dragonfly",
+        target_os = "fuchsia",
+        target_os = "illumos",
+        target_os = "netbsd",
+        target_os = "openbsd",
+        target_os = "redox",
+        target_os = "solaris",
+    )))]
+    pub fn set_recv_tclass_v6(&self, recv_tclass: bool) -> io::Result<()> {
+        let recv_tclass = i32::from(recv_tclass);
```
Comment (Thomasdezeeuw) [inline src/socket.rs:1732]: Not needed, `bool as c_int` works.
```suggestion
```

**Review by Thomasdezeeuw — COMMENTED** [review]
Can you switch the ordering of the functions, first the getter than the setter.

**Comment by duskmoon314** [comment]
Done

Should I remove `let recv_tos = i32::from(recv_tos);` too?

**Comment by Thomasdezeeuw** [comment]
> Should I remove `let recv_tos = i32::from(recv_tos);` too?

Sure 👍

**Review by Thomasdezeeuw — APPROVED** [review]

**Comment by Thomasdezeeuw** [comment]
@duskmoon314 can you fix the failing CI for Windows? See https://github.com/rust-lang/socket2/actions/runs/3740167945/jobs/6348219101

**Comment by duskmoon314** [comment]
Sure. I am looking for MS documentation on this setting

**Comment by duskmoon314** [comment]
It seems that Microsoft doesn't allow modifying the traffic class of IPv6 win-socket as in Unix-like OS.

For IPv4, using `IP_TOS` is allowed and may correspond to history. And the doc notes that QoS API is recommended:

> Do not use. Type of Service (TOS) settings should only be set using the Quality of Service API. See [Differentiated Services](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/qos/differentiated-services) in the Quality of Service section of the Platform SDK for more information.
> 
> https://learn.microsoft.com/en-us/windows/win32/winsock/ipproto-ip-socket-options

And for IPv6, no `IPV6_TCLASS` is mentioned in the doc. So this is probably why the test fails since I assume it is usable and just not be mentioned.

I am considering adding `target_os = windows` to the cfg `not` attribute. Any suggestions?

**Comment by Thomasdezeeuw** [comment]
> I am considering adding `target_os = windows` to the cfg `not` attribute. Any suggestions?

Yes, that's the way to go I think. However if it's not available on Windows it need to be behind the `all` feature and we'll likely want to move it to `sys/unix.rs`. For an example see https://github.com/rust-lang/socket2/blob/fa67e04d6fdb3dca0075db2a88acbffe39980eee/src/sys/unix.rs#L1048-L1076

**Comment by duskmoon314** [comment]
Sure. I will modify the code ASAP.

File: src/sys/unix.rs [hunk 4]
```diff
@@ -76,6 +76,30 @@ pub(crate) use libc::{
...
+        target_os = "dragonfly",
+        target_os = "freebsd",
+        target_os = "fuchsia",
+        target_os = "linux",
+        target_os = "macos",
+        target_os = "netbsd",
+        target_os = "openbsd"
+    )
+))]
+pub(crate) use libc::IPV6_TCLASS;
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:102]: No longer needed.

```suggestion
```

File: src/sys/unix.rs [hunk 5]
```diff
@@ -2057,6 +2081,84 @@ impl crate::Socket {
...
+                target_os = "linux",
+                target_os = "macos",
+                target_os = "netbsd",
+                target_os = "openbsd"
+            )
+        )))
+    )]
+    pub fn tclass_v6(&self) -> io::Result<u32> {
+        unsafe {
+            getsockopt::<c_int>(self.as_raw(), IPPROTO_IPV6, IPV6_TCLASS)
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:2121]:
```suggestion
            getsockopt::<c_int>(self.as_raw(), IPPROTO_IPV6, libc::IPV6_TCLASS)
```

File: src/sys/unix.rs [hunk 6]
```diff
@@ -2057,6 +2081,84 @@ impl crate::Socket {
...
+                target_os = "fuchsia",
+                target_os = "linux",
+                target_os = "macos",
+                target_os = "netbsd",
+                target_os = "openbsd"
+            )
+        )))
+    )]
+    pub fn set_tclass_v6(&self, tclass: u32) -> io::Result<()> {
+        unsafe { setsockopt(self.as_raw(), IPPROTO_IPV6, IPV6_TCLASS, tclass as c_int) }
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:2160]:
```suggestion
        unsafe { setsockopt(self.as_raw(), IPPROTO_IPV6, libc::IPV6_TCLASS, tclass as c_int) }
```

**Review by Thomasdezeeuw — APPROVED** [review]

**Comment by Thomasdezeeuw** [comment]
I'll squash the commit I've added.

**Comment by Thomasdezeeuw** [comment]
Thanks @duskmoon314!

**Pull request merged**